games/hangman.c

```diff
@@ -0,0 +1,291 @@
+                }
+            }
+
+            printf("\nYou have %d guesses left.", (12 - game.incorrect));
+            printf("\nPlease enter a letter: ");
+            scanf(" %c", &guess);
+            guess = tolower(guess);
+
+        } while (new_guess(guess, game.guesses, game.guesses_size) != -1);
+
+        game.guesses[game.guesses_size] = guess; // adds new letter to guesses array
+        game.guesses_size++; // updates size of guesses array
+
+        if (in_word(guess, game.current_word, game.size) == 1) {
+            printf("That letter is in the word!");
+            for (int i = 0; i < game.size; i++) {
+                if ((game.current_word[i]) == guess) {
+                    game.hidden[i] = guess;
+                }
+            }
+        } else {
+            printf("That letter is not in the word.\n");
+            (game.incorrect)++;
+        }
+        picture(game.incorrect);
+    }
+
+    won(game.current_word, game.incorrect);
+    return 0;
+}
+
+/**
+ * @brief checks if letter has been guessed before
+ * @param new_guess letter that has been guessed by player
+ * @param guesses array of player's previous guesses
+ * @param size size of guesses[] array
+ * @returns 1 if letter has been guessed before
+ * @returns -1 if letter has not been guessed before
+ */
+int new_guess(char new_guess, const char guesses[], int size) {
+
+    for (int j = 0; j < size; j++) {
+        if (guesses[j] == new_guess) {
+            printf("\nYou have already guessed that letter.");
+            return 1;
+        }
+    }
+
+    return -1;
+}
+
+/**
+ * @brief checks if letter is in current word
+ * @param letter letter guessed by player
+ * @param word current word
+ * @param size length of word
+ * @returns 1 if letter is in word
+ * @returns -1 if letter is not in word
+ */
+int in_word(char letter, const char word[], unsigned int size) {
+
+    for (int i = 0; i < size; i++) {
+        if ((word[i]) == letter) {
+            return 1;
+        }
+    }
+
+    return -1;
+}
+
+/**
+ * @brief creates a new game - generates a random word and stores in global variable current_word
+ * @returns current_game - a new game instance containing randomly selected word, its length and hidden version of word
+ */
+struct game_instance new_game() {
+
+    char word[30]; // used throughout function
+
+    FILE *fptr;
+    fptr = fopen("games/words.txt", "r");
+
+    if (fptr == NULL){
+        fprintf(stderr, "File not found.\n");
+        exit(EXIT_FAILURE);
+    }
+
+    // counts number of words in file - assumes each word on new line
+    int line_number = 0;
+    while (fgets(word, 30, fptr) != NULL) {
+        line_number++;
+    }
+
+    rewind(fptr);
+
+    // generates random number
+    int random_num;
+    srand(time(NULL));
+    random_num = rand() % line_number;
+
+    // selects randomly generated word
+    int s = 0;
+    while (s <= random_num){
+        fgets(word, 30, fptr);
```

Comment:
Reading through the word list twice struck me as rather inefficient the first time I looked at your code, but I didn't raise it at the time because of other more important issues. Caching is an obvious solution, and we can do it reasonably efficiently with, say, a linked list, but why bother complicate things with industrial grade code?